plugins/filter_kubernetes/kubernetes.c

```diff
@@ -335,14 +342,49 @@ static int pack_map_content(msgpack_packer *pck, msgpack_sbuffer *sbuf,
         if (log_index == i &&
             (merge_status == MERGE_UNESCAPED || merge_status == MERGE_PARSED)) {
             if (merge_status == MERGE_UNESCAPED || ctx->keep_log == FLB_TRUE) {
-                msgpack_pack_object(pck, k);
+                msgpack_pack_object(pck, *k);
                 msgpack_pack_str(pck, ctx->unesc_buf_len);
                 msgpack_pack_str_body(pck, ctx->unesc_buf, ctx->unesc_buf_len);
             }
         }
         else { /* MERGE_BINARY ? */
-            msgpack_pack_object(pck, k);
-            msgpack_pack_object(pck, v);
+            msgpack_pack_object(pck, *k);
+            msgpack_pack_object(pck, *v);
+        }
+
+        if (k->type == MSGPACK_OBJECT_BIN) {
+            key_ptr  = (char *) k->via.bin.ptr;
+            key_size = k->via.bin.size;
+        }
+        else if (k->type == MSGPACK_OBJECT_STR) {
+            key_ptr  = (char *) k->via.str.ptr;
+            key_size = k->via.str.size;
+        }
+
+        if (key_size < (sizeof(buf_key) - 1)) {
+            memcpy(buf_key, key_ptr, key_size);
+            buf_key[key_size] = '\0';
+            ptr_key = buf_key;
+        }
+        else {
+            /* Long map keys have a performance penalty */
+            ptr_key = flb_malloc(key_size + 1);
+            if (!ptr_key) {
+                flb_errno();
+                return -1;
+            }
+
+            memcpy(ptr_key, key_ptr, key_size);
+            ptr_key[key_size] = '\0';
+        }
+
+        if (ctx->replace_dots == FLB_TRUE) {
+          char *p   = ptr_key;
+          char *end = ptr_key + key_size;
+          while (p != end) {
+            if (*p == '.') *p = '_';
+            p++;
+          }
```

Comment:
I don't see where ptr_key is used, do I miss something?

Reply:
Maybe I can remove `p` variable and use `ptr_key` directly. Is that OK ?

Reply:
My point is that I don't see where you make use of the modified string: you free it just after building it. What are you trying to achieve? What use case do you want to address?

Reply:
I'm trying to do the same as in the [Elasticsearch output plugin](https://github.com/fluent/fluent-bit/blob/b99898292f42669c5023c286bb41dd0653a7764e/plugins/out_es/es.c#L90). There is a piece of code that allow to change `.` character to `_` in all keys of each log message.

But (as you can see :sweat_smile: ) I'm really not a C developer (nor a developer at all actually) so I just tried to copy the ES plugin, to help implementing this feature on Kubernetes filter. Maybe this is not that simple :smile:

Reply:
My question was more about the final (functional) use case than the technical use case (replacing dots from keys in Kubernetes filter). Why do you need to replace dots and why do you need to do it in the Kubernetes filter?

Reply:
The Kubernetes filter collects annotations and labels from Kubernetes API and add them as keys/value to the log message object. Kubernetes annotations keys usually contains `.` (like `nginx.ingress.kubernetes.io/app-root` or `service.beta.kubernetes.io/aws-load-balancer-proxy-protocol`, that kind of keys).

Those dots in the key name can be annoying for some systems (like Elasticsearch) so I propose to add an option to remove them (like [Elastic Beats did](https://github.com/elastic/beats/issues/9860). I understand that the Elasticsearch output of fluentbit already have this *dedot* feature, but it is common to have a a bunch of Fluentbit instances pushing to a Fluentd instance pushing to a store (like Elasticsearch). And in this case, I think it could be interesting to manage this "*dots issue*" at the beginning of the log pipeline, so the Kubernetes filter.

This is just a feature suggestion to solve an #1134, I'll understand if this feature is not accepted since there are other workarounds.

Reply:
In my opinion, replacing dots should better be done in Fluentd Elasticsearch output plugin. But if you want to do it in Fluent Bit Kubernetes filter plugin, K8s labels and annotations are packed at line 398. Here you're acting on the map which comes from the input (for example tail input plugin) after it has been potentially parsed. Labels and annotations aren't in this map and are packed later (beginning at line 398) ;-)

Reply:
![Facepalm gif](https://media.giphy.com/media/ADr35Z4TvATIc/giphy.gif)

Thanks for the advice, I really tried to implement this feature brainless...
I'll try to put all of this during the step of packing labels and annotations. But maybe I should wait for a validation of the feature on FluentBit. Maybe it should not be implemented, I don't know.